compiler/plugins/target/ROCM/ROCMTarget.cpp

```diff
@@ -372,44 +359,63 @@ class ROCMTargetBackend final : public TargetBackend {
       // multi-threading issues.
       llvm::LLVMContext context;
 
-      auto llvmModule =
-          mlir::translateModuleToLLVMIR(innerModuleOp, context, libraryName);
-      if (!llvmModule) {
-        return variantOp.emitError() << "failed to translate the MLIR LLVM "
-                                        "dialect to the native llvm::Module";
-      }
-
+      // Set up attributes so upstream's conversions work right.
+      auto *rocdlDialect =
+          innerModuleOp.getContext()->getLoadedDialect<ROCDL::ROCDLDialect>();
+      auto kernelAttrHelper = rocdlDialect->getKernelAttrHelper();
+      auto flatWgAttrHelper = rocdlDialect->getFlatWorkGroupSizeAttrHelper();
+      auto reqdWgSizeAttrHelper =
+          rocdlDialect->getReqdWorkGroupSizeAttrHelper();
+      auto wavesPerEqHelper = rocdlDialect->getWavesPerEuAttrHelper();
+      auto unitAttr = executableBuilder.getUnitAttr();
       for (auto func : innerModuleOp.getOps<LLVM::LLVMFuncOp>()) {
-        int32_t flatWgSize = 1;
-        llvm::Function *llvmFunc = llvmModule->getFunction(func.getName());
-        if (llvmFunc->isDeclaration())
+        // Un-exported functions are library functions or otherwise not
+        // not kernels.
+        if (!exportOpMap.contains(func.getName()))
           continue;
+        kernelAttrHelper.setAttr(func, unitAttr);
         auto exportOp = exportOpMap[func.getName()];
         if (auto workgroupSizeAttr = exportOp.getWorkgroupSize()) {
+          SmallVector<int32_t> wgSizes;
+          wgSizes.reserve(3);
+          int32_t flatWgSize = 1;
           for (Attribute attr : *workgroupSizeAttr) {
-            flatWgSize *= cast<IntegerAttr>(attr).getInt();
+            int32_t val = cast<IntegerAttr>(attr).getInt();
+            flatWgSize *= val;
+            wgSizes.push_back(val);
           }
+          reqdWgSizeAttrHelper.setAttr(
+              func, executableBuilder.getDenseI32ArrayAttr(wgSizes));
+          flatWgAttrHelper.setAttr(
+              func, executableBuilder.getStringAttr(Twine(flatWgSize) + "," +
+                                                    Twine(flatWgSize)));
```

Comment:
This function is getting very long; could we move this into a helper function?
By 'this' I mean most of the loop body.